### ISSUE TITLE
Server get wrong segment download url from zookeeper

### ISSUE THREAD

**Issue by code4love** [linkedin/pinot#13]
We found that pinot-server get wrong segment download url from zookeeper, like http://localhost:9999/segments/myTable/mySegment_0.tar.gz. It seems that controller set "localhost" as it's hostname.
```
[pool-2-thread-1] 2015/06/19 12:36:58.665 INFO [com.linkedin.pinot.common.utils.FileUploadUtils] [] Get file from:http://localhost:9999/segments/myTable/mySegment_0.tar.gz to local file:/tmp/PinotSegment8098/segmentTar/mySegment_0.tar.gz
[pool-2-thread-1] 2015/06/19 12:36:58.666 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[pool-2-thread-1] 2015/06/19 12:36:58.666 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] Retrying request
[pool-2-thread-1] 2015/06/19 12:36:58.667 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[pool-2-thread-1] 2015/06/19 12:36:58.667 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] Retrying request
[pool-2-thread-1] 2015/06/19 12:36:58.667 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[pool-2-thread-1] 2015/06/19 12:36:58.667 INFO [org.apache.commons.httpclient.HttpMethodDirector] [] Retrying request
[pool-2-thread-1] 2015/06/19 12:36:58.667 ERROR [com.linkedin.pinot.common.utils.FileUploadUtils] [] Caught exception
java.net.ConnectException: Connection refused
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:345)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at java.net.Socket.connect(Socket.java:538)
        at java.net.Socket.<init>(Socket.java:434)
        at java.net.Socket.<init>(Socket.java:286)
        at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:80)
        at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:122)
        at org.apache.commons.httpclient.HttpConnection.open(HttpConnection.java:707)
        at org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.java:387)
        at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:171)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:397)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:323)
at com.linkedin.pinot.common.utils.FileUploadUtils.getFile(FileUploadUtils.java:90)
        at com.linkedin.pinot.server.starter.helix.SegmentOnlineOfflineStateModelFactory$SegmentOnlineOfflineStateModel.downloadSegmentToLocal(SegmentOnlineOfflineStateModelFactory.java:327)
        at com.linkedin.pinot.server.starter.helix.SegmentOnlineOfflineStateModelFactory$SegmentOnlineOfflineStateModel.onBecomeOnlineFromOfflineForOfflineSegment(SegmentOnlineOfflineStateModelFactory.java:212)
        at com.linkedin.pinot.server.starter.helix.SegmentOnlineOfflineStateModelFactory$SegmentOnlineOfflineStateModel.onBecomeOnlineFromOffline(SegmentOnlineOfflineStateModelFactory.java:128)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.apache.helix.messaging.handling.HelixStateTransitionHandler.invoke(HelixStateTransitionHandler.java:344)
        at org.apache.helix.messaging.handling.HelixStateTransitionHandler.handleMessage(HelixStateTransitionHandler.java:290)
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:93)
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:50)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

PS: In com.linkedin.pinot.tools.admin.command.StartControllerCommand, you can find following codes:
```
 conf.setControllerHost("localhost");
 conf.setControllerVipHost("localhost");
```

I think it should be 
```
String hostname = InetAddress.getLocalHost().getHostName()
```

**Comment by dongyanfeng** [comment]
The same problems！！！？

**Comment by mayankshriv** [comment]
As mentioned in the pinot-admin wiki, pinot-admin currently works on local host. Pinot Broker/Server/Controller should be started on local host using pinot-admin. It will have cluster mode support in future. Hope this clarifies.

**Comment by kishoreg** [comment]
The instructions provided should work for cluster mode as well. I think the
problem is that
https://github.com/linkedin/pinot/blob/b565c3d1cbebc03d0c1e97750a19fb617344675b/pinot-tools/src/main/java/com/linkedin/pinot/tools/admin/command/StartControllerCommand.java
execute method that is hardcoding controllerHost to localhost.

conf.setControllerHost("localhost");
changing this to conf.setControllerHost(
netAddress.getLocalHost().getHostName()); should fix it. Can you give it a
try and let us know?. Feel free to submit a patch

thanks,
Kishore G

**Comment by mayankshriv** [comment]
The issue with hardcoded "localhost" has now been resolved.

**Comment by dongyanfeng** [comment]
After you solved the problem, unable to create table, please confirm

**Comment by mayankshriv** [comment]
@dongyanfeng what is the issue you see with creating table?

**Comment by code4love** [comment]
This issue had resolved  in commit c875cc0
